### DISCUSSION TITLE
How to use spherical coordinates? (Or how to make custom cells given a set of points/vertices?)

### DISCUSSION THREAD

**Discussion by jagerber48** [pyvista/pyvista#4914] [Q&A] [answered]
I am working on a spherically symmetric *volume* problem (visualization of atomic orbitals). For this problem (visualizing an atomic orbital) I have a function `psi` of 3D space. It is easy enough for me to calculate the value of this function in either cartesian or spherical coordinates. The functions I am visualizing have spherical symmetry. This means that to get high resolution volumetric or surface contour visualizations it is most efficient to sample the function in spherical coordinates. This is because if I sample the space using cartesian coordinates the sampling rate along a surface of constant radius is highest along the x/y/z axes, but drops down at e.g. 45 degrees between any of these cardinal directions. FUTHERMORE, in the radial direction, these functions are more rapidly spatially varying near the origin and then vary less slowly as the radial distance increases. This means it is efficient to sample the radius non-linearly. Again, I can accomplish a similar effect using cartesian coordinates with, e.g. a structured grid, but again, this doesn't sample uniformly along the polar/azimuthal directions.

So with all of that out of the way, what I am needing is a mesh whose vertices are along spherical coordinates (possibly with the radial coordinate sampled non-linearly). It is easy for me to create numpy arrays of `(x, y, z)` coordinates which place the vertices at the appropriate positions. But what I don't know how to do is take these `(x, y, z)` coordinates and create a pyvista mesh out of them whose cells span the obvious coordinate cells and, e.g. "wrap around" at `theta=pi` and `phi=2 pi`.

Here is example code:
```python
import numpy as np
import pyvista as pv

r_1d = np.linspace(0.1, 10, 5)
theta_1d = np.linspace(0, np.pi, 25)
phi_1d = np.linspace(0, 2*np.pi, 25)

r, theta, phi = np.meshgrid(r_1d, theta_1d, phi_1d, indexing='ij')

x = r * np.sin(theta) * np.cos(phi)
y = r * np.sin(theta) * np.sin(phi)
z = r * np.cos(theta)

xyz = np.transpose(
    np.stack(
        [x.ravel(order='F'), y.ravel(order='F'), z.ravel(order='F')],
        axis=0
    )
)

pl = pv.Plotter(notebook=False)

mesh = pv.PointSet(xyz)

pl.add_mesh(mesh)
pl.add_points(mesh.points)
pl.show()
```

This code produces a point set that looks like this:
![image](https://github.com/pyvista/pyvista/assets/48464524/102587fe-a487-43bb-9721-a590a2b3f3a2)

The question is how to take this same point set and add in cells in each coordinate pseudo-cubiod?

One note: In this example code I don't include the origin. That is because at the origin the topology changes if a vertex is included at the origin. Basically the coordinates become degenerate and the origin vertex would be included in MANY cells whereas typical non-origin vertices are as-a-rule included in exactly 8 cells. I would be happy to see answers that either do or don't include the origin in the mesh.

**Comment by MatthewFlamm** [comment, marked as answer]
I believe you are so close to answering your question.  The key here is that Structured grid only needs to have an i-j-k referencing system for the cells, not that the domain itself has to be rectangular. I think we should add this example to the documentation for `StructuredGrid`.

```python
import numpy as np
import pyvista as pv

r_1d = np.linspace(0.1, 10, 5)
theta_1d = np.linspace(0, np.pi, 25)
phi_1d = np.linspace(0, 2*np.pi, 25)

r, theta, phi = np.meshgrid(r_1d, theta_1d, phi_1d, indexing='ij')

x = r * np.sin(theta) * np.cos(phi)
y = r * np.sin(theta) * np.sin(phi)
z = r * np.cos(theta)

mesh = pv.StructuredGrid(x, y, z)

pl = pv.Plotter(notebook=False)
pl.add_mesh(mesh, style='wireframe')
pl.add_mesh(mesh.slice(), show_edges=True)
pl.show()
```

**Reply by MatthewFlamm** [reply]
With a lot more effort, you can also construct an UnstructuredGrid that goes all the way to the center.  The reason you would need an UnstructuredGrid is that you would have mixed cell types, e.g. hexahedron for most of the domain and ~tetrahedron~ pyramids to the center point.

**Reply by MatthewFlamm** [reply]
I will also note that there is an example related to atomic orbitals  https://docs.pyvista.org/version/stable/examples/99-advanced/atomic-orbitals.html, but it takes a different approach.  It uses `ImageData` which uniformly samples space in a cartesian sense.

**Reply by jagerber48** [reply]
Yes! this works! I figured out the `StructuredGrid` approach a little after writing the question. A couple things.

- One concern I have is that when I plot the `StructuredGrid` just as a mesh (using default options for `add_mesh`, similar to your example) I get an outer spherical shell (corresponding to the largest radius) and an inner spherical shell (corresponding to the smallest radius) and these two are connected by a thin meridional slice at `phi=0`. This made me worried there is something defective about the geometry.
- Also, my code seems to work even if I do allow the radius to go all the way down to 0 and this makes some contour plots look better. Don't know if I should be surprised that it just works or not.
- Thank you for sharing the atomic orbitals example! I didn't know that was there. I am working on something very similar. I'll post a picture below, maybe I could add to the existing example page. Importantly, working in spherical coordinates and sampling non-linearly makes it much easier to plot higher order orbitals while maintaining good resolution and speed.
- Finally, I am encountering issues when I try to do volume plotting using an RGBA dataset on the `StructuredGrid`. This is true even if the `StructuredGrid` is sampled on a cartesian grid. The colors and opacity are all wacky. However, the rendering looks fine if I use an equivalent `RectilinearGrid` or view the `StructuredGrid` in paraview (but in this case the rendering is slow). This is probably fodder for a separate issue. If so I'll post more details in a separate question.

Here's an example of a contour plot of an atomic orbital I'm able to make using non-linearly sampled spherical coordinates. This is for `(n, l, m) = (8, 4, 1)`. This is a complex orbital where the hue encodes the wavefunction phase and this contour encloses 50% of the wavefunction probability. r, theta, phi are sampled each with 150 points. r is sampled according to a `sinh` function so that points near the origin are sampled more finely than further out (where you can see the wavefunction lobes get bigger). an `xyz` corner has been clipped/clamped out. This rendered in a couple seconds on my laptop and the interactive view is totally responsive.

![image](https://github.com/pyvista/pyvista/assets/48464524/23ea033a-27e2-44cc-b2aa-2d783f40ef85)

**Reply by MatthewFlamm** [reply]
Very cool!  If you use StructuredGrid all the way to 0, I suspect you end up with degenerate hexahedra, i.e. four points are all at the origin.  This might be fine for what you need, but degenerate cells might not always behave the way you expect.

**Reply by MatthewFlamm** [reply]
Pinging @adeak for whether it makes sense to add this approach to the example.

**Reply by MatthewFlamm** [reply]
Also, please open another issue/discussion on the rendering issue.

**Reply by adeak** [reply]
Very nice stuff @jagerber48 and thanks for the ping @MatthewFlamm.

I think such a grid based approach (as opposed to volumes) would be great to have in the atomic orbitals example. _And_ it would be great to have a curvilinear example in the `StructuredGrid` docs examples.

**Reply by jagerber48** [reply]
Thanks both!

@MatthewFlamm 
- Makes sense what you say about the degenerate hexahedra cells.
- Any comment about my first concern above about the mesh rendering of the `StructuredGrid` with the meridional plane? Is that some defect in the geometry or something not to be worried about?
- Ok, I'll open an issue about the volume rendering issue.

@adeak Thanks! I'll look into bundling my code into a small example! Yes, the curvilinear `StructuredGrid` example would be nice.

**Reply by MatthewFlamm** [reply]
> Any comment about my first concern above about the mesh rendering of the StructuredGrid with the meridional plane? Is that some defect in the geometry or something not to be worried about?

This is because `StructuredGrid` inherently uses an i-j-k ordering to define the cells.  This means that there is a defined beginning and end to the mesh in each i-j-k direction.  This results in the seam that you see; it is both the beginning and end of your mesh.  You might see extrapolation artifacts there due to extrapolation vs. interpolation.  You can make an `UnstructuredGrid` instead which would allow you to define points and the cell structure within those points.  This would allow you to have no seam, or a seam if you prefer.  I am pretty sure that `vtkExplicitStructuredGrid` might also allow no seam, but I'm not confident that PyVista's `ExplicitStructuredGrid` can be used this way.

**Reply by MatthewFlamm** [reply]
https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.ExplicitStructuredGrid.html#pyvista-explicitstructuredgrid
> Can be initialized by the following:
> - Creating an empty grid
> - From a vtk.vtkExplicitStructuredGrid or vtk.vtkUnstructuredGrid object
> - From a VTU or VTK file
> - From dims and corners arrays

This defininition based on dims and corner arrays seems like a mistake to me for ExplicitStructuredGrid as it precludes the option to define a more flexible grid.  I will open an issue for it.

**Reply by jagerber48** [reply]
@MatthewFlamm ok interesting. So I think for my specific application with atomic orbitals I've seen that `StructuredGrid` can work well and easily. But I think if I REALLY wanted spatial accuracy then the way to go would be an `UnstructuredGrid`. I'll probably look into constructing this for fun. My strategy would maybe be something like: create a `StructuredGrid` that stops one radial step before the origin, one azimuthal step before the origin meridian and one polar step away from the North and South poles. This would give me all of the points I need except all the points along the polar axis (including the origin). These polar axis points could easily be added into the point set by hand. Finishing it off would be a matter of adding in all of the cells missing in the radial, polar, and azimuthal gaps near the respective origins.

Alternatively, it may just be easier to loop out radially creating shells of points and cells as you. The first solid sphere including the origin will be a set of pyramids I think and then every shells moving outwards will follow the same routine adding on hexahedrons. It may be easier to keep track of which point is point for constructing the cells in this fashion.

**Reply by MatthewFlamm** [reply]
You can create an `UnstructuredGrid` directly from the `StructuredGrid` using `cast_to_unstructured_grid`, if this is an easier starting point.  However, I suspect you will get a better result with only a small amount of extra work by starting from scratch.  Even when you start with a partially created grid, you will need to create routines to find the relevant points for your additions, and these routines will contain a significant fraction of the code for a from-scratch implementation.

You might still find it useful to look at the internal 'structure' of the `UnstructuredGrid` converted from the `StructuredGrid`.  In particular pay attention to the `cells` attribute and the `celltypes` attribute. Starting from scratch would also allow you to more easily avoid degenerate Hexahedra not only at the origin, but also the ones that connect to the axis of rotation.

**Reply by jagerber48** [reply]
@MatthewFlamm Yeah, I've been looking at this. It is interesting. If I start with a `StructuredGrid` that includes the origin as well as points along the polar axis then I end up with duplicated points and some degenerate hexahedra. 

- Away from the origin along the polar axis the hexahedra should be wedges. 
- Near the origin along the polar axis the hexahedra should be tetrahedra
- Near the origin away from the polar axis the hexehedra should be pyramids.

One easy approach I've made progress on is to take the `StructuredGrid`, convert it to an `UnstructuredGrid` then call `clean()`. This eliminates the duplicated points and connects the cells appropriately. Then you can `triangulate()` this mesh to turn all the hexahedra (including some degenerate ones) into tetrahedra. You are still left with some cells that are degenerate tetrahedra, but these are pretty easy to detect and can just be deleted from the mesh I think.

I haven't tried out the above strategy to completion but it should work. One possible downside of this approach is that it results in all tetrahedra when you could be using hexahedra, wedges, pyramids, and tetrahedra. I guess the downside is that there are more cells than you technically need which would be a performance hit (but with increased resolution) if you are doing cell-based operations or visualizations.

Another approach would be to manually convert the degenerate hexahedra into wedges, pyramids, and tetrahedra accordingly. This is pretty easy (ignoring cell orientation, not sure if that matters for 3D applications, I know some 2D algorithms that rely on cell orientation) for tetrahedra (just eliminate vertex repetitions and convert cell type) and pyramids (delete all repeated vertices and add the vertex to the end of the cell) but converting hexahedra to wedges seems to involve a large number of cases that need to be handled. The switch case wouldn't be too complicated, just tedious. Is there code available that can analyze a mesh and fix vertex degenerate 3D cells by replacing them with a non-vertex-degenerate 3D cell of a different type or delete the cell entirely if it degenerate to the point of being 0D 1D or 2D?

I think these comments are still technically relevant to the original question I asked so unless I hear otherwise I'll keep posting these comments here. If I get code working I may add it as another answer, specifically addressing the part of the question about how to get a "clean" solid mesh with points along spherical coordinates.

**Reply by MatthewFlamm** [reply]
I think that once this is implemented, it should also be in the library.  We have `Sphere` which generates a `PolyData`, but it would be generally useful to have `FilledSphere`, or something like this, which is `UnstructuredGrid`.

**Reply by MatthewFlamm** [reply]
In vtk 9.3.0 there will be a `vtkCleanUnstructuredGridCells` algorithm that says it handles degenerate cells.  There may be hope that this will simplify operations when it is easier to generate degenerate cells.  https://vtk.org/doc/nightly/html//classvtkCleanUnstructuredGridCells.html

Under investigation, it seems to remove degenerate cells, so this doesn't work.

**Comment by jagerber48** [comment]
Here is one routine to get a clean solid mesh with vertices at polar coordinates that goes all the way down to the origin. We can get most of the way there using the `meshgrid` from the OP.

```python
import numpy as np

r_1d = np.linspace(0, 10, 10)
theta_1d = np.linspace(0, np.pi, 10)
phi_1d = np.linspace(0, 2*np.pi, 10)

r, theta, phi = np.meshgrid(r_1d, theta_1d, phi_1d, indexing='ij')

structured_mesh = pv.StructuredGrid(x, y, z)
```

This creates a mesh with vertices and cells in the right spots. This is best thought of as starting with a 3D cartesian grid and then bending and warping that grid until it is in a spherical shape. There are 2 problems. The first is that many cells overlap (for example there will 100 vertices all located at the origin, one for each `theta`/`phi` combination). The second is that all of the cells in a `StructuredGrid` are hexahedra with 8 vertices and 6 faces. But because of the overlapping, many of the cells in the spherical structured grid become degenerate, meaning 2 or more of their points collapse into one, thus reducing the number of faces.

One strategy to clean up these degeneracies is as follows.

First we cast the `StructuredGrid` to an `UnstructuredGrid`. We must give up some of the regularity of the `StructuredGrid` because all of our cells are not hexahedra.

```python
unstructured_mesh = structured_mesh.cast_to_unstructured_grid()
```

Next we eliminate duplicate vertices using `clean()`

```python
cleaned_mesh = unstructured_mesh.clean(tolerance=1e-4)
```

This method deletes overlapping vertices and re-assigns the corresponding vertices in the connected cells. We've gotten rid of degenerate vertices, but we still need to fix degenerate hexahedra cells. Inspection reveals that there are 4 types of hexahedra in this unstructured mesh.

- Away from the origin and away from the polar axis there are regular non-degenerate hexahedra left over from the original structured grid.
- Away from the origin but near the polar axis the hexahedra have collapsed into wedges running up and down the polar axis.
- Near the origin but away from the polar axis we see that one entire face of each hexahedron has collapsed into the origin point. This leaves us with pyramids.
- Near the origin and near the polar axis our cells have collapsed down to tetrahedra circling the polar axis.

Ideally we would take the degenerate hexahedra, cut out the extra degenerate vertices, and convert them to the appropriate cell type. But doing this is not entirely trivial since you need to handle many possible cases for which specific vertices in the hexahedron cell are degenerate. Below I'll take an easier approach.

Instead of worrying about manually converting cell types, we will just convert all cells to collections of tetrahedra using triangulation.

```python
tri_mesh = cleaned_mesh.triangulate()
```

This triangulated mesh STILL has some degenerate cells. But a degenerate tetrahedron is just a 2D surface or 1D line. These degenerate cells can be entirely eliminated from the mesh without affecting the volume. We use these helpers methods to help us delete the degenerate cells:

```python
def cells_to_cell_list(cells):
    cells = cells.copy()
    cell_list = list()
    while len(cells):
        cell_len = cells[0]
        cell = cells[1:cell_len + 1]
        cell_list.append(cell)
        cells = cells[cell_len+1:]
    return cell_list

def cell_list_to_cells(cell_list):
    cells = list()
    for cell in cell_list:
        cell_len = len(cell)
        cells.append(cell_len)
        cells += list(cell)
    return np.array(cells)

def delete_degenerate_cells(mesh):
    points = mesh.points
    cells = mesh.cells
    celltypes = mesh.celltypes
    cell_list = cells_to_cell_list(cells)
    
    cleaned_cell_list = list()
    cleaned_celltypes = list()
    for cell, celltype in zip(*(cell_list, celltypes)):
        if len(cell) == len(set(cell)):
            cleaned_cell_list.append(cell)
            cleaned_celltypes.append(celltype)
    cleaned_cells = cell_list_to_cells(cleaned_cell_list)
    cleaned_celltypes = np.array(cleaned_celltypes)

    fixed_mesh = pv.UnstructuredGrid(cleaned_cells, cleaned_celltypes, points)
    return fixed_mesh
```
The linear `cells` object is inconvenient to work with so we cast it to a `cell_list` object which just lists the cells. The first two methods just allow us to switch back and forth between `cells` and `cell_list`. `delete_degenerate_cells` loops through the list of cells and finds out if a cell is degenerate by converting it to a set and checking if its length has changed. A new mesh is returned constructed from the remaining non-degenerate cells.

We can visualize this with some code to identify and extract degenerate cells:

```python
def get_degenerate_cell_indices(mesh):
    cells = mesh.cells.copy()
    cell_list = cells_to_cell_list(cells)
    degenerate_cell_indices = list()
    for cell_idx, cell in enumerate(cell_list):
        if len(cell) != len(set(cell)):
            degenerate_cell_indices.append(cell_idx)
    return degenerate_cell_indices

def plot_wireframe_with_degenerate_cells(mesh):
    degenerate_cell_indices = get_degenerate_cell_indices(mesh)

    pl = pv.Plotter(notebook=False)
    pl.add_mesh(mesh.points, color='black', opacity=0.2)
    pl.add_mesh(mesh.extract_all_edges(), color='blue', opacity=0.2)
    if len(degenerate_cell_indices) > 0:
        pl.add_mesh(mesh.extract_cells(degenerate_cell_indices).extract_all_edges(), color='red', opacity=1)
    else:
        print('No degenerate cells')
    pl.show()
```

Calling 
```python
plot_wireframe_with_degenerate_cells(tri_mesh)
```

produces
![image](https://github.com/pyvista/pyvista/assets/48464524/c4e5795c-c7ff-4643-b871-6f476d7bfce8)
We see the edges of degenerate tetrahedra left over near the origin and polar axis.

Calling 
```python
plot_wireframe_with_degenerate_cells(fixed_mesh)
```

produces
![image](https://github.com/pyvista/pyvista/assets/48464524/7566c8b0-6014-4527-8966-cdcfcaa7c892)
It is the same mesh but there are no longer any degenerate edges.

This mesh could be a useful starting point to do calculations involving functions with spherical-like symmetry.

Note: I think it would be preferable to use the strategy of casting degenerate cells to the appropriate new cell type rather than casting everything to tetrahedra. There are two downsides to the triangulation approach. The first is cosmetic. The wireframe doesn't *look* like a spherical coordinate grid the way the (degenerate) structured grid version does. The second issue is performance. The triangulated mesh has many more cells so cell-based calculations and visualizations will take more time. Though there will be higher spatial resolution. 

The degenerate cell conversion algorithm could get pretty complicated though. A special case conversion for this problem could be written with a little bit of tedium for this problem since we are only dealing with hexahedra being converted to wedges, pyramids, and tetrahedra. But a general conversion (which should probably exist?) would involve checking what type of degeneracy has occurred, figuring out if a higher vertex count cell has been converted to a lower vertex count cell, OR if the degeneracy actually just needs to be resolved by triangulation, and then returning the appropriate new cells.

**Reply by MatthewFlamm** [reply]
Here is code to make a symmetric, filled spherical mesh.  Also with an example of how to adjust the sampling density in the radial direction.  This could use a lot of cleanup, e.g. making a reusable function for indexing:

![image](https://github.com/pyvista/pyvista/assets/39341281/03520caf-c783-4576-9408-32c25581d531)



```py
import numpy as np
import pyvista as pv

nr=5
ntheta=5
nphi=8


def spherical_to_cartesian(r, theta, phi):
    # takes in 1D ndarrays or floats for each, returns (n,3) points in cartesian
    r, theta, phi = np.meshgrid(r, theta, phi, indexing='ij')

    x = r * np.sin(theta) * np.cos(phi)
    y = r * np.sin(theta) * np.sin(phi)
    z = r * np.cos(theta)

    return np.vstack((x.ravel(), y.ravel(), z.ravel())).transpose()

#r_1d = np.insert(np.geomspace(0.1, 10, nr-1), 0, 0.0)
r_1d = np.linspace(0, 10, nr)
theta_1d = np.linspace(0, np.pi, ntheta)
phi_1d = np.linspace(0, 2*np.pi, nphi)

#points on axis
points = [[0., 0., 0.]]
points.extend(spherical_to_cartesian(r_1d[1:], theta_1d[0], phi_1d[0]))
points.extend(spherical_to_cartesian(r_1d[1:], theta_1d[-1], phi_1d[0]))

# rest of points with phi changing quickest
for ir in r_1d[1:]:
    for itheta in theta_1d[1:-1]:
        points.extend(spherical_to_cartesian(ir, itheta, phi_1d[:-1]))

cells = []
celltypes = []

 # 2*nr-1 points on axis
 # At each r, theta, nphi-1 points
 # At each r, ntheta-2 rings of points: theta[0] and theta[-1] are on the axis

# first make the tetras that form with origin and axis point
#   origin is 0
#   first axis point is 1
#   other points at 2*nr-1 + (0:nphi-1)

for iphi in range(nphi-1):
    start = 2*nr-1
    upper_range = (1 + iphi) % (nphi - 1)
    cells.append(4)
    cells.extend([0, 1, start+iphi, start+upper_range])

    celltypes.append(pv.CellType.TETRA)

# next tetras that form with origin and bottom axis point
# 2 theta points are already on axis and we want the last remaning theta level, so (ntheta-3)
# other points at 2*nr-1 + (ntheta-3)*(nphi-1) + (0:nphi-1)

for iphi in range(nphi-1):
    start = 2*nr - 1 + (ntheta - 3) * (nphi - 1)
    upper_range = (1 + iphi) % (nphi - 1)
    cells.append(4)
    cells.extend([0, nr, start+iphi, start+upper_range])

    celltypes.append(pv.CellType.TETRA)

# Pyramids that form to origin but without an axis point
# start with the same points as above, but connect to next theta points

for itheta in range(ntheta-3):
    for iphi in range(nphi-1):
        start = 2*nr - 1 + itheta*(nphi-1)
        upper_range = (1 + iphi) % (nphi - 1)
        next = 2*nr - 1 + (itheta+1)*(nphi-1)
        cells.append(5)
        cells.extend([start+iphi, start+upper_range, next+upper_range, next+iphi, 0])

        celltypes.append(pv.CellType.PYRAMID)

# Wedges form between two r levels at first and last theta position
#   At each r level, the triangle is formed with axis point,  two phi positions
# First go upwards
for ir in range(nr-2):
    axis0 = ir + 1
    axis1 = ir + 2
    for iphi in range(nphi-1):
        start = 2*nr-1 + ir*(nphi-1)*(ntheta-2)
        upper_range = (1 + iphi) % (nphi - 1)

        next = 2*nr-1 + (ir+1)*(nphi-1)*(ntheta-2)
        cells.append(6)
        cells.extend([axis0, start+iphi, start+upper_range, axis1, next+iphi, next+upper_range])

        celltypes.append(pv.CellType.WEDGE) 

# now go downwards
for ir in range(nr-2):
    axis0 = nr + ir
    axis1 = nr + ir + 1
    for iphi in range(nphi-1):
        start = 2*nr-1 + ir*(nphi-1)*(ntheta-2) + (ntheta - 3) * (nphi - 1)
        upper_range = (1 + iphi) % (nphi - 1)

        next = 2*nr-1 + (ir+1)*(nphi-1)*(ntheta-2) + (ntheta - 3) * (nphi - 1)
        cells.append(6)
        cells.extend([axis0, start+iphi, start+upper_range, axis1, next+iphi, next+upper_range])

        celltypes.append(pv.CellType.WEDGE) 


# Form Hexahedra
# Hexahedra form between two r levels and two theta levels and two phi levels
#   Order by r levels
for ir in range(nr-2):
    for itheta in range(ntheta-3):
        for iphi in range(nphi-1):
            start0 = 2*nr-1 + ir*(nphi-1)*(ntheta-2) + itheta*(nphi - 1)
            upper_range = (1 + iphi) % (nphi - 1)

            next0 = 2*nr-1 + ir*(nphi-1)*(ntheta-2)  + (itheta + 1)*(nphi - 1)
            start1 = 2*nr-1 + (ir+1)*(nphi-1)*(ntheta-2) + itheta*(nphi - 1)
            next1 = 2*nr-1 + (ir+1)*(nphi-1)*(ntheta-2)  + (itheta + 1)*(nphi - 1)


            cells.append(8)
            cells.extend([start0+iphi, start0+upper_range, next0+upper_range, next0+iphi, start1+iphi, start1+upper_range, next1+upper_range, next1+iphi])

            celltypes.append(pv.CellType.HEXAHEDRON) 

mesh = pv.UnstructuredGrid(cells, celltypes, points)

mesh_cell_center = mesh.cell_centers()
mesh["radial_position"] = np.linalg.norm(mesh_cell_center.points, axis=1)

pl = pv.Plotter()
pl.add_mesh(mesh.explode(1.25))
pl.show()
```

**Reply by jagerber48** [reply]
Beautiful! This is what I wanted to do but I wasn't able to (or didn't want to because it sounded tricky) sit down and wrap my head around how to specify/keep track of the vertex indices when trying to construct all the different "special cases" for the types of cells. I'm curious to test this out and see if it is at least as performant as the `StructuredGrid` with degeneracy for making atom wavefunction contour plots.

**Reply by MatthewFlamm** [reply]
Tracking this request to include in base library #4943

**Reply by jagerber48** [reply]
One more vague comment I want to include for completion of my understanding: It is exactly the "special caseness" of the different "types" of vertices and cells within the solid spherical mesh that we need to use an `UnstructuredGrid` and not a `StructuredGrid`. This is exactly the use case for `UnstructuredGrid`. If there weren't special cases then a `StructuredGrid` would suffice. Before working on this problem this distinction wasn't clear to me. I think this is a nice example to go in the documentation explaining why one might ever want an `UnstructuredGrid`.

**Reply by MatthewFlamm** [reply]
`UnstructuredGrid` is the most general type of dataset that can hold any arbitrary cell type in either 2D or 3D.  Before we get to `StructuredGrid`, take `ImageData`.  It can only hold a `Voxel` cell type and the cells and points are arranged in a cartesian-aligned i-j-k ordering with regular spacing.  This dataset type does not explicitly store any point positions or cell connectivity, but is computed on-the-fly if needed.  `RectilinearGrid` similar but allows for nonregular spacing.  `StructuredGrid` does not impose the cartesian-alignment on the points, and thus now explicitly stores the point positions. It still enforces that the cell connectivity is i-j-k.  Now that these are non-cartesian, the cells are `Hexahedra`.

Since `UnstructuredGrid` can hold datasets of any arbitrary complexity including cell shape (Voxel, Hexahedra, Tetrahedra, nonlinear cells, Polyhedra, etc.), connectivity of cells, mixed cell types and arbitrary point locations.  Any other grid can be converted to it, but not vice versa.  But this has downsides as the more compact representation datasets may be more performant for _some_ operations and they definitely consume less memory.  It is also harder to manually construct due to the flexibility inherent in its structure.

This is touched on extremely briefly here https://docs.pyvista.org/version/stable/user-guide/data_model.html#the-pyvista-dataset.  vtk also has some description of this.  See this (although this is through Paraview, the concepts are the same) https://docs.paraview.org/en/latest/UsersGuide/understandingData.html

If you have any ideas on how to improve the documentation, please contribute, even if it is a proposal for what type of information should go where.
